### DISCUSSION TITLE
Google Authentication always redirecting to localhost:3000

### DISCUSSION THREAD

**Discussion by Minisaints** [vercel/community#2076] [Help] [answered]
### Summary

Hello,

I have my NextJS project hosted on Vercel and it works great. However, logging in using Google Auth, I am redirected back to localhost:3000 instead of "https://***.vercel.app".

I have specified the Environment variable "NEXTAUTH_URL" in Vercel settings to "https://***.vercel.app", redeployed and also added in the redirect URI list in Google Dev Console. It's like Vercel is ignoring the "NEXTAUTH_URL" environment variable.

What am I missing here? 

Thanks

### Additional information

```
export const authOptions: NextAuthOptions = {
    adapter: UpstashRedisAdapter(db),
    session: {
        strategy: "jwt"
    },
    pages: {
        signIn: '/login'
    },
    providers: [
        GoogleProvider({
            clientId: getGoogleCredentials().clientId,
            clientSecret: getGoogleCredentials().clientSecret
        })
    ],
    callbacks: {
        async jwt({ token, user }) {
            const dbUserResult = (await fetchRedis("get", `user:${token.id}`)) as string | null
            if (!dbUserResult) {
                token.id = user!.id
                return token;
            }

            const dbUser = JSON.parse(dbUserResult)

            return {
                id: dbUser.id,
                name: dbUser.name,
                email: dbUser.email,
                picture: dbUser.image
            }
        },
        async session({ session, token }) {
            if (token) {
                session.user.id = token.id,
                    session.user.name = token.name,
                    session.user.email = token.email,
                    session.user.image = token.picture
            }

            return session;
        },
        redirect() {
            return '/dashboard'
        }
    },
    jwt: {
        secret: process.env.NEXTAUTH_JWT_SECRET
    },
    secret: process.env.NEXTAUTH_SECRET
}
```

**Comment by amyegan** [comment]
Hi @Minisaints! The `VERCEL_URL` environment variable is used when [deploying to Vercel](https://next-auth.js.org/deployment#vercel), so `NEXTAUTH_URL` is not needed.

The typical provider config uses `process.env` variables like
```
GoogleProvider({
  clientId: process.env.GOOGLE_ID,
  clientSecret: process.env.GOOGLE_SECRET,
}),
```
So the first thing I recommend to verify that the Google ID and secret you have set up are associated with the production URL rather than localhost. Then, you should check that `getGoogleCredentials()` is configured correctly and getting the expected values.

If that makes no difference, you should reach out to the NextAuth folks who will have more insight into what may be going wrong. They have their own [NextAuth Discussions](https://github.com/nextauthjs/next-auth/discussions) page where you can reach out for help.

**Reply by craftogrammer** [reply]
Same here, I have setup the project in similar way. I can confirm everything is working on netlify, render.com, but only Vercel is ignoring the NEXTAUTH_URL even if it's not needed, as I have setup all ENV vars in all platforms, and I have checked it many times. The problem is specific to Vercel only.

I can share more info if you need.

**Reply by Minisaints** [reply]
Hi Amy,

The Google ID/Secret are all accessible and SSO is functioning great locally in development. It's when I deploy the project to Vercel, the issue appears. I have added in the VERCEL_URL into environment variables within Vercel with no luck as well.

I'm going to re-create the Vercel project step by step again and see if that resolves it. 

Thanks

**Reply by Minisaints** [reply]
I have built a separate project that only allows Google Auth. These are the steps taken. (Everything worked great locally too).

1. Created a new NextJS 13 project and installed Vercel CLI and Next-Auth.
2. Populated .env with Google ID/Secret and Vercel URL (VERCEL_URL) ready for deployment.
3. Implemented Google Auth into the project.
4. Used Vercel CLI to link the project and deploy.
5. Imported environment variables into Vercel using the .env file.
6. Re-deployed and the latest deployment is below.

Going to the deployment below allows you to select your Google Credential but on redirect, you land on localhost:3000.

https://vercel-njs-auth.vercel.app/auth/signin

**Reply by craftogrammer** [reply]
> I have built a separate project that only allows Google Auth. These are the steps taken. (Everything worked great locally too).
> 
> 1. Created a new NextJS 13 project and installed Vercel CLI and Next-Auth.
> 2. Populated .env with Google ID/Secret and Vercel URL (VERCEL_URL) ready for deployment.
> 3. Implemented Google Auth into the project.
> 4. Used Vercel CLI to link the project and deploy.
> 5. Imported environment variables into Vercel using the .env file.
> 6. Re-deployed and the latest deployment is below.
> 
> Going to the deployment below allows you to select your Google Credential but on redirect, you land on localhost:3000.
> 
> https://vercel-njs-auth.vercel.app/auth/signin

And, if you host the same project on non-vercel platforms, it works.

**Reply by Minisaints** [reply]
I've not deployed it to another platform but running locally using Node/NJS works fine.

**Reply by amyegan** [reply, marked as answer]
I'm trying to reproduce this problem, but it's all working as expected for me. I deployed my own copy of the [starter code](https://github.com/nextauthjs/next-auth-example) with Google Auth and it redirected back to my example site. The [NextAuth example site](https://next-auth-example.vercel.app/) uses Next.js v13 and it's also redirecting back correctly.

I would recommend triple checking that your `NEXTAUTH_URL` and Google Cloud OAuth settings are in agreement. You need to make sure you have set `https://{YOUR_DOMAIN}/api/auth/callback/google` as one of Google OAuth client ID's `Authorized redirect URIs` as described in the [NextAuth Google configuration doc](https://next-auth.js.org/providers/google). 

If you're still having trouble, please reach out to the [NextAuth maintainers](https://github.com/nextauthjs/next-auth/discussions) or [Google Cloud Customer Care](https://next-auth.js.org/providers/google) for more specific with code and configuration issues.

**Reply by Minisaints** [reply]
Hi Amy,

Thanks for clarifying that.

Google OAuth redirect API, NEXTAUTH_URL and the VERCEL_URL all match. The weird thing is that if I change all three to localhost:5000, It redirects me; it's just not allowing me to redirect using the Vercel production URL.

![image](https://user-images.githubusercontent.com/34611622/233738400-0567bd3a-8f11-406f-b7b3-ecf084f0778e.png)

I wasn't aware of the starter code from NextAuth so I will wire that up, push to Vercel and see if that works. If not, I might find the nearest cliff!

Thanks again.

Craig
